<div style="text-align: center; padding: 20px; font-family: Vazir;">
<h1 align="center" style="font-size: 28px; color:rgb(64, 244, 202); width: 100%;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>تمرین ۲<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1>
<h2 style="color:rgb(90, 255, 184); font-size: 20px;">Logistic Regression and Naive Bayes</h2>
<p align="center" style="color: #666; font-size: 16px;">فرهاد نصری - علیرضا زمانی</p>
<p align="center" style="color: #666; font-size: 16px; margin-bottom: 30px;">farhadnasri999@gmail.com - shigzv@gmail.com</p>

<div dir="rtl" style="border: 2px dashed rgb(90, 255, 184); border-radius: 8px; padding: 20px; margin: 20px auto; max-width: 500px; text-align: right;">
<p style="color: rgb(64, 244, 202); font-size: 18px; margin-bottom: 15px;">📝 مشخصات دانشجو:</p>
<p style="color: #666; margin: 5px;">نام و نام خانوادگی: امین آقاکثیری</p>
<p style="color: #666; margin: 5px;">شماره دانشجویی: 810101381</p>
<p style="color: #666; margin: 5px;">تاریخ ارسال: 1404/8/24</p>
</div>
</div>

<div dir="rtl" style="text-align: justify; padding: 25px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<div style="padding-right:100px">
📋 <b>ساختار تمرین:</b>
<li><b>سوال اول - <span dir="ltr">Logistic Regression and Naive Bayes (from scratch)</span> (60)</b></li>
<ul>
<li>بخش اول: پیش پردازش داده‌های متنی و استفاده از Bag of Words</li>
<li>بخش دوم: جداسازی داده‌های آموزش و آزمایش</li>
<li>بخش سوم: پیاده‌سازی معیار‌های ارزیابی</li>
<li>بخش چهارم: پیاده‌سازی Logistic Regression</li>
<li>بخش پنجم: پیاده‌سازی Naive Bayes</li>
<li>بخش ششم: تحلیل نتایج</li>
</ul>
<li><b>سوال دوم - <span dir="ltr">Logistic Regression and Naive Bayes (e.g. 
sklearn)</span> (40)</b></li>
<ul>
<li>بخش اول: استخراج ویژگی‌های ساختاری</li>
<li>بخش دوم: استخراج ویژگی‌های آماری و محتوایی</li>
<li>بخش سوم: استخراج ویژگی‌های دلخواه</li>
<li>بخش چهارم: ترکیب همه ویژگی‌های استخراج‌شده</li>
</ul>
</div>
</div>
<div dir='rtl' style="line-height: 1.8; font-family: Vazir; font-size: 16px; margin-top: 20px; background-color: #e8eaf6; padding: 15px; border-radius: 8px; color:black">
💡 <b>نکات مهم:</b>
<br>
در سوال دو مجاز به استفاده از کتابخانه‌های آماده برای مدل‌ها و متریک‌ها هستید ولی در سوال یک، همان‌طور که در متن سوال هم ذکر شده، باید از پایه پیاده‌سازی کنید.
</div>
</div>

# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>سوال اول: پیاده‌سازی Logistic Regression و Naive Bayes <br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">
در این سوال، شما باید این دو مدل پایه‌ای برای طبقه‌بندی متون را از صفر (بدون استفاده از کتابخانه‌های آماده) پیاده‌سازی کنید. هم‌چنین نیاز است، پیش‌پردازش داده‌ها و ارزیابی عملکرد مدل‌ها را نیز انجام دهید.




## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش اول: پیش پردازش داده‌های متنی و استفاده از Bag of Words</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
در گام پیش‌پردازش، هدف ما آماده‌سازی داده‌های متنی برای استفاده در مدل‌های یادگیری ماشین است. از آن‌جا که مدل‌ها با داده‌های عددی کار می‌کنند، باید متن خام را به شکلی عددی تبدیل کنیم تا بتوانند الگوهای موجود در واژه‌ها را بیاموزند. در این بخش، قصد داریم با تبدیل متون به نمایش Bag of Words، هر ایمیل را بر اساس تعداد تکرار واژه‌هایش به یک بردار عددی تبدیل کنیم.
<br>
دیتاست emails_1.csv شامل دو ستون است:
<br>
text (متن ایمیل)
<br>
status (برچسب، مشخص‌کننده‌ی spam یا ham بودن ایمیل)
<br>
شما باید مراحل زیر را انجام دهید:
<br>
۱. ستون status را به مقادیر عددی تبدیل کنید (spam → 1 و ham → 0).
<br>
۲. متن‌ها را پاک‌سازی کنید: تمام کاراکترهای غیر الفبایی (اعداد، علائم و غیره) را حذف و همه‌ی حروف را کوچک (lowercase) کنید.
<br>
۳. با استفاده از روش Bag of Words، فراوانی کلمات را محاسبه کنید و فقط ۱۵ کلمه‌ی پرتکرار را نگه دارید.
<br>
💡 نکته: برای این بخش می‌توانید از CountVectorizer در کتابخانه‌ی sklearn استفاده کنید.
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
یک دیتافریم جدید که شامل ۱۵ ویژگی + ستون status است. (تمام ۱۰ ردیف این دیتافریم را نمایش دهید.)
</p>

In [239]:
import pandas as pd
import re
from sklearn.feature_extraction.text import CountVectorizer
import numpy as np

In [240]:
df = pd.read_csv('./datasets/q1/emails_1.csv')
df['status'] = df['status'].map({'ham': 0, 'spam': 1})

In [241]:
df.head()


text  status
0  🎉 Congratulations! You have won a FREE ticket ...       1
1  Reminder: The project meeting will be held tom...       0
2  Exclusive offer just for YOU! Get a free-trial...       1
3  Hey team - the report looks great; please revi...       0
4  You’ve been selected for a *FREE* gift card! C...       1

In [242]:
def clean_text(text):
    text = text.lower()                          
    text = re.sub(r'[^a-z\s]', '', text)          
    text = re.sub(r'\s+', ' ', text).strip()     
    return text

df["clean_text"] = df["text"].apply(clean_text)
df.head()

text  status  \
0  🎉 Congratulations! You have won a FREE ticket ...       1   
1  Reminder: The project meeting will be held tom...       0   
2  Exclusive offer just for YOU! Get a free-trial...       1   
3  Hey team - the report looks great; please revi...       0   
4  You’ve been selected for a *FREE* gift card! C...       1   

                                          clean_text  
0  congratulations you have won a free ticket to ...  
1  reminder the project meeting will be held tomo...  
2  exclusive offer just for you get a freetrial o...  
3  hey team the report looks great please review ...  
4  youve been selected for a free gift card claim...

In [243]:
vectorizer = CountVectorizer(max_features=15)
counts = vectorizer.fit_transform(df["clean_text"]).toarray()
vocab = vectorizer.get_feature_names_out()
new_df = pd.DataFrame(counts, columns=vocab)
new_df['status'] = df['status']
new_df.head()

and  at  by  claim  enjoy  for  free  get  our  please  project  the  to  \
0    0   0   0      1      0    0     2    0    1       0        0    0   1   
1    1   1   0      0      0    0     0    0    0       1        2    1   0   
2    0   0   0      0      1    2     1    1    1       0        0    0   0   
3    1   0   1      0      0    0     0    0    0       1        0    1   0   
4    0   0   0      1      1    1     1    0    0       0        0    0   0   

   you  your  status  
0    1     1       1  
1    0     1       0  
2    1     0       1  
3    0     1       0  
4    0     2       1

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
از این مرحله به بعد، دیگر با فایل emails_1.csv و دیتافریم حاصل از آن کاری نخواهیم داشت. برای سهولت کار، فایل emails_2.csv در اختیار شما قرار گرفته است. این فایل، نتیجه‌ی همان فرایند پیش‌پردازش بر روی ۵۱۷۲ ایمیل واقعی است. به بیان ساده‌تر، در این مجموعه داده، هر ردیف نمایانگر یک ایمیل و هر ستون نشان‌دهنده‌ی فراوانی یکی از ۳۰۰۰ کلمه‌ی پرتکرار در کل داده‌ها است.
در ادامه، مدل‌هایی که پیاده‌سازی می‌کنید باید روی این مجموعه داده، آموزش داده شوند. بنابراین، در این بخش فایل emails_2.csv را بخوانید و محتوای آن را نمایش دهید تا با ساختار داده آشنا شوید.
</p>



<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
پنج ردیف اول دیتافریم مذکور را نمایش دهید.
</p>

In [277]:
df = pd.read_csv('./datasets/q1/emails_2.csv')
df.head()

Email No.  the  to  ect  and  for  of    a  you  hou  ...  connevey  jay  \
0   Email 1    0   0    1    0    0   0    2    0    0  ...         0    0   
1   Email 2    8  13   24    6    6   2  102    1   27  ...         0    0   
2   Email 3    0   0    1    0    0   0    8    0    0  ...         0    0   
3   Email 4    0   5   22    0    5   1   51    2   10  ...         0    0   
4   Email 5    7   6   17    1    5   2   57    0    9  ...         0    0   

   valued  lay  infrastructure  military  allowing  ff  dry  Prediction  
0       0    0               0         0         0   0    0           0  
1       0    0               0         0         0   1    0           0  
2       0    0               0         0         0   0    0           0  
3       0    0               0         0         0   0    0           0  
4       0    0               0         0         0   1    0           0  

[5 rows x 3002 columns]

## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش دوم: جداسازی داده‌های آموزش و آزمایش</div>

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
داده‌ها را با نسبت ۸۰ درصد برای آموزش و ۲۰ درصد برای آزمایش تقسیم کنید. در انتها تعداد نمونه‌ها را چاپ کنید.
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
X_train, X_test, y_train, y_test<br>
تعداد نمونه‌های X_train و X_test 
</p>

In [278]:
from sklearn.model_selection import train_test_split
X = df.drop('Prediction', axis=1)
y = df['Prediction']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=30)
X_train.reset_index(drop=True, inplace=True)
X_test.reset_index(drop=True, inplace=True)
y_train.reset_index(drop=True, inplace=True)
y_test.reset_index(drop=True, inplace=True)

In [279]:
print(f"X_train head:\n{X_train.head()}")

X_train head:
    Email No.  the  to  ect  and  for  of   a  you  hou  ...  enhancements  \
0  Email 1714    3   0    1    0    0   4  12    1    1  ...             0   
1  Email 4228    0   1    1    1    0   0   5    0    0  ...             0   
2  Email 2627    0   2    1    0    0   1  33    0    0  ...             0   
3   Email 862    0   1    1    0    0   1   0    0    0  ...             0   
4  Email 2564    5   2    2    1    3   3  46    4    0  ...             0   

   connevey  jay  valued  lay  infrastructure  military  allowing  ff  dry  
0         0    0       0    0               0         0         0   2    0  
1         0    0       0    0               0         0         0   0    0  
2         0    0       0    0               0         0         0   2    0  
3         0    0       0    0               0         0         0   0    0  
4         0    0       0    0               0         0         0   0    0  

[5 rows x 3001 columns]


In [280]:
y_train

0       0
1       0
2       1
3       1
4       0
       ..
4132    0
4133    1
4134    1
4135    0
4136    1
Name: Prediction, Length: 4137, dtype: int64

In [281]:
print(f"X_train number of samples: {X_train.shape[0]}")

X_train number of samples: 4137


In [282]:
print(f"X_test number of samples: {X_test.shape[0]}")

X_test number of samples: 1035


In [283]:
_, counts = np.unique(y_train, return_counts=True)
print("Number of hams and spams in y_train:")
print("hams:", counts[0])
print("spams:", counts[1])

Number of hams and spams in y_train:
hams: 2937
spams: 1200


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش سوم: پیاده‌سازی معیار‌های ارزیابی</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
پیش از آن‌که وارد مرحله‌ی پیاده‌سازی و آموزش مدل‌های طبقه‌بندی شویم، لازم است معیارهایی برای سنجش عملکرد آن‌ها تعریف کنیم. این معیارها به ما کمک می‌کنند تا بفهمیم مدل تا چه اندازه در تشخیص درست نمونه‌های مثبت و منفی موفق عمل کرده است.
<br>
در این بخش، توابع زیر را برای حالت دودویی (binary classification) پیاده‌سازی کنید:
<br>
accuracy(y_true, y_pred) — نسبت پیش‌بینی‌های درست به کل نمونه‌ها
<br>
precision(y_true, y_pred) — درصد پیش‌بینی‌های مثبت که واقعاً مثبت بوده‌اند
<br>
recall(y_true, y_pred) — درصد نمونه‌های مثبت واقعی که مدل آن‌ها را درست شناسایی کرده است
<br>
f1_score(y_true, y_pred) — میانگین هارمونیک بین precision و recall، برای ایجاد توازن میان آن دو
<br>
هر تابع باید مقدار عددی متناظر با معیار مورد نظر را برگرداند.
<br>
 به توابع بالا دو ورودی زیر را بدهید و خروجی بگیرید:    y_true = [0, 1, 1, 0, 1] ---- y_pred = [0, 1, 0, 0, 1]
<br>
💡 نکته: برای این بخش می‌توانید از  کتابخانه‌ی numpy استفاده کنید.
</p>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
مثال:
<br>
y_true = [0, 1, 1, 0, 1]
<br>
y_pred = [0, 1, 0, 0, 1]
<br>
print(accuracy(y_true, y_pred))  # 0.80
<br>
print(precision(y_true, y_pred)) # 1.00
<br>
print(recall(y_true, y_pred))    # 0.66
<br>
print(f1_score(y_true, y_pred))  # 0.80
<br>
</p>

In [8]:
import numpy as np
def get_accuracy(y_true, y_pred):
    correct_predictions = np.sum(y_true == y_pred)
    total_predictions = len(y_true)
    return correct_predictions / total_predictions
def get_precision(y_true, y_pred):
    tp = np.sum((y_true == 1) & (y_pred == 1))
    predicted_pos = np.sum(y_pred == 1)
    return tp / predicted_pos if predicted_pos > 0 else 0
def get_recall(y_true, y_pred):
    tp = np.sum((y_true == 1) & (y_pred == 1))
    actual_pos = np.sum(y_true == 1)
    return tp / actual_pos if actual_pos > 0 else 0
def f1_score(y_true, y_pred):
    precision = get_precision(y_true, y_pred)
    recall = get_recall(y_true, y_pred)
    return (2* (precision * recall)) / (precision + recall) if (precision + recall) > 0 else 0

In [286]:
y_true = np.array([0,1,1,0,1])
y_pred = np.array([0,1,0,0,1])
print(f"Accuracy: {get_accuracy(y_true, y_pred):.2f}")
print(f"Precision: {get_precision(y_true, y_pred):.2f}")
print(f"Recall: {get_recall(y_true, y_pred):.2f}")
print(f"F1 Score: {f1_score(y_true, y_pred):.2f}")

Accuracy: 0.80
Precision: 1.00
Recall: 0.67
F1 Score: 0.80


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش چهارم: پیاده‌سازی Logistic Regression</div>

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
کلاس Logistic Regression را به طور کامل پیاده سازی کنید.سپس مدل را بر روی داده‌های train آموزش دهید و در انتها دقت مدل را باتوجه به متریک های تعریف شده بر روی داده‌های test گزارش کنید.
<br>
💡استفاده از هایپر پارامتر‌های مناسب مانند نرخ یادگیری و تعداد دوره‌های آموزش بر عهده خودتان است.
<br>
💡میتوانید قبل از آموزش داده ها را نرمال سازی کنید.
<br>
💡مجاز به استفاده از کتابخانه اماده نیستید.
</p>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
معیار‌های پیاده سازی شده را بر روی داده‌های test خروجی بگیرید و نتایج را چاپ کنید
</p>

In [284]:
class LogRegression:
    def __init__(self, learning_rate, epochs):
        self.learning_rate = learning_rate
        self.epochs = epochs
        self.weights = None
        self.bias = None

    def sigmoid(self, inp):
        # preventing overflow
        inp = np.clip(inp, -200, 200)
        return 1 / (1 + np.exp(-inp))

    def fit(self, X, y):
        smaples_num, features_num = X.shape
        self.weights = np.zeros(features_num)
        self.bias = 0
        for epoch_num in range(self.epochs):
            # shuffle samples at each epoch for SGD
            indices = np.arange(smaples_num)
            np.random.shuffle(indices)
            for i in indices:
                out = np.dot(X[i], self.weights) + self.bias
                y_pred = self.sigmoid(out)
                dw = (y_pred - y[i]) * X[i]
                db = y_pred - y[i]
                self.weights -= self.learning_rate * dw
                self.bias -= self.learning_rate * db
            if (epoch_num+1) % 5 == 0:
                loss = self.get_loss(X, y)
                print(f"Epoch {epoch_num+1}, Loss: {loss:.3f}")
            
    
    def get_loss(self,X, y):
        y_pred = self.sigmoid(np.dot(X, self.weights) + self.bias)
        return -np.mean(y*np.log(y_pred+1e-10) + (1-y)*np.log(1-y_pred+1e-10))

    def predict(self, X):
        out = np.dot(X, self.weights) + self.bias
        outputs = self.sigmoid(out)
        y_pred = [1 if i > 0.5 else 0 for i in outputs]
        return np.array(y_pred)

In [287]:
# Standardize features
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train.iloc[:,1:].values)
X_test_scaled  = scaler.transform(X_test.iloc[:,1:].values)

In [288]:
logreg = LogRegression(learning_rate=0.01, epochs=100)
logreg.fit(X_train_scaled, y_train.values)
y_pred = logreg.predict(X_test_scaled)
# X_train_arr = X_train.iloc[:,1:].values
# X_test_arr = X_test.iloc[:,1:].values
# logreg.fit(X_train_arr, y_train.values)
# y_pred = logreg.predict(X_test_arr)
print(f"\nAccuracy: {get_accuracy(y_test.values, y_pred):.3f}")
print(f"Precision: {get_precision(y_test.values, y_pred):.3}")
print(f"Recall: {get_recall(y_test.values, y_pred):.3}")
print(f"F1 Score: {f1_score(y_test.values, y_pred):.3}")

Epoch 5, Loss: 0.015
Epoch 10, Loss: 0.009
Epoch 15, Loss: 0.007
Epoch 20, Loss: 0.006
Epoch 25, Loss: 0.006
Epoch 30, Loss: 0.005
Epoch 35, Loss: 0.005
Epoch 40, Loss: 0.004
Epoch 45, Loss: 0.004
Epoch 50, Loss: 0.004
Epoch 55, Loss: 0.004
Epoch 60, Loss: 0.003
Epoch 65, Loss: 0.003
Epoch 70, Loss: 0.003
Epoch 75, Loss: 0.003
Epoch 80, Loss: 0.003
Epoch 85, Loss: 0.003
Epoch 90, Loss: 0.003
Epoch 95, Loss: 0.002
Epoch 100, Loss: 0.002

Accuracy: 0.964
Precision: 0.91
Recall: 0.973
F1 Score: 0.94


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش پنجم: پیاده‌سازی Naive Bayes</div>

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
کلاس Multinomial Naive Bayes را به طور کامل پیاده سازی کنید. سپس مدل را بر روی داده‌های train آموزش دهید و در انتها دقت مدل را باتوجه به معیار های تعریف شده بر روی داده‌های test گزارش کنید.
<br>
- چرا از مدل‌های دیگر مانند Gaussian Naive Bayes استفاده نکردیم؟ آیا از مدل‌های دیگر Naive Bayes می‌توان استفاده کرد؟
<br>
💡مجاز به استفاده از کتابخانه اماده نیستید.
<br>
💡برای آشنایی با Naive Bayes می‌توانید به Appendix کتاب jurafsky مراجعه کنید.
</p>

<div dir="rtl" style="background:#fffbe6;font-family: Vazir;border:1px dashed #f0ad4e;padding:12px;border-radius:8px;color:#111;line-height:1.9;overflow:auto;max-width:100%;box-sizing:border-box;">
✍️ <b>پاسخ تشریحی بخش پنجم:</b><br>
در اینجا، داده‌ها به‌صورت Bag of Words (BoW) پیش‌پردازش شده‌اند. یعنی هر ایمیل به مجموعه‌ای از ویژگی‌ها تبدیل شده که نشان می‌دهد هر کلمه چند بار در آن متن تکرار شده است. بنابراین ویژگی‌های ما عددی، گسسته، غیرپیوسته و شامل مقادیر صفر زیاد هستند.

به همین دلیل از مدل‌هایی مانند Gaussian Naive Bayes استفاده نکردیم، چون Gaussian NB فرض می‌کند ویژگی‌ها از یک توزیع نرمال (گوسی) پیروی می‌کنند. این فرض در داده‌های متنی درست نیست، چون فراوانی کلمات معمولا نرمال نیست، بلکه پراکندگی زیادی دارد و بیشتر مقادیر آن صفر هستند. در نتیجه، استفاده از Gaussian NB برای داده‌های متنی معمولا باعث کاهش دقت مدل می‌شود و تخمین‌های درستی از احتمال‌ها به ما نمی‌دهد. اگر از بردار های پیوسته مانند embedding ها استفاده می‌کردیم، آن‌گاه Gaussian NB می‌توانست گزینه مناسبی باشد. همچنین در صورتی که داده ما پیوسته بود، می‌توانستیم از Gaussian NB استفاده کنیم.

در مقابل، مدل‌هایی از خانواده‌ Naive Bayes وجود دارند که دقیقا برای داده‌های متنی طراحی شده‌اند. بهترین گزینه در این حالت Multinomial Naive Bayes است، چون فرض آن بر پایه‌ی شمارش کلمات است و می‌تواند به‌خوبی الگوهای مربوط به تعداد تکرار هر واژه در کلاس‌های مختلف (مثل spam یا ham) را یاد بگیرد.

مدل دیگری هم به نام Bernoulli Naive Bayes وجود دارد که زمانی استفاده می‌شود که ویژگی‌ها فقط نشان دهند آیا یک کلمه در متن وجود دارد یا نه (یعنی مقادیر صفر و یک). اما چون در داده‌ی فعلی ما تعداد واقعی تکرار کلمات در نظر گرفته شده و داده‌ها باینری نیستند، استفاده از Bernoulli NB در این حالت مناسب نیست. همچنین اگر توزیع کلاس ها نامتوازن باشد، استفاده از Complement Naive Bayes هم می تواند مفید باشد و در واقع کاری که می کند این است که به‌جای اینکه توزیع کلمات یک کلاس کوچک را از خود آن کلاس ضعیف یاد بگیریم
از تمام داده‌های دیگر (که بسیار بزرگ‌تر و پایدارتر هستند) استفاده می‌ کند. اما در اینجا چون کلاس هایمان نامتوازن نیست، بهتر است از همان Multinomial Naive Bayes استفاده کنیم.

در نتیجه، برای داده‌ی فعلی که بر پایه‌ Bag of Words است، Multinomial Naive Bayes بهترین و منطقی‌ترین انتخاب است و مدل‌هایی مانند Gaussian NB به‌دلیل ناسازگاری با نوع داده، استفاده نمی‌شوند.
</div>



<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
معیار‌های پیاده سازی شده را بر روی داده‌های test خروجی بگیرید و نتایج را چاپ کنید
</p>

In [259]:
class MultinomialNaiveBayes:
    def __init__(self):
        self.class_probs = None # P(C_k)
        self.word_probs = None # P(x_i | C_k)
        self.classes = None              
    
    def fit(self, X, y):
        smaples_num, features_num = X.shape
        self.classes = np.unique(y)        
        classes_num = len(self.classes)
        self.class_probs = np.zeros(classes_num)
        self.word_probs = np.zeros((classes_num, features_num))
        
        for i in range(len(self.classes)):
            c = self.classes[i]
            X_class = X[y == c]
            self.class_probs[i] = X_class.shape[0] / smaples_num
            counts = np.sum(X_class, axis=0)
            self.word_probs[i, :] = (counts + 1) / (np.sum(counts) + features_num)

    
    # I used logs for calculating probs to prevent underflowing (n_i is the number of x_i word occurunce)
    # log(P(C)*product(P(x_i|C)^n_i)) = log(P(C)) + sum(n_i*log(P(x_i|C)))
    def get_log_probs(self, X):
        smaples_num = X.shape[0]
        classes_num = len(self.classes)
        log_probs = np.zeros((smaples_num, classes_num))
        for i in range(classes_num):
            # log(P(C_k))
            class_prob = np.log(self.class_probs[i])
            # log(P(X | C_k)) = sum over features of n_i * log(P(x_i | C_k))
            sum_word_probs = np.dot(X, np.log(self.word_probs[i, :]))
            log_probs[:, i] = class_prob + sum_word_probs  
        return log_probs
    
    def predict(self, X):
        log_probs = self.get_log_probs(X)
        return np.argmax(log_probs, axis=1)


In [260]:
naive_bayes = MultinomialNaiveBayes() 
X_train_arr = X_train.iloc[:,1:].values
X_test_arr = X_test.iloc[:,1:].values
naive_bayes.fit(X_train_arr, y_train.values)
y_pred = naive_bayes.predict(X_test_arr)
print(f"Accuracy: {get_accuracy(y_test.values, y_pred):.3f}")
print(f"Precision: {get_precision(y_test.values, y_pred):.3f}")
print(f"Recall: {get_recall(y_test.values, y_pred):.3f}")
print(f"F1 Score: {f1_score(y_test.values, y_pred):.3f}")

Accuracy: 0.954
Precision: 0.899
Recall: 0.947
F1 Score: 0.922


## <div style="text-align: center; direction: rtl; font-family: Vazir;"> بخش ششم: تحلیل نتایج</div>

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
نتایج دو مدل پیاده سازی شده را با یکدیگر مقایسه کنید و تحلیلی بر نتایج داشته باشید.<br>
    
</p>

<div dir="rtl" style="background:#fffbe6;font-family: Vazir;border:1px dashed #f0ad4e;padding:12px;border-radius:8px;color:#111;line-height:1.9;overflow:auto;max-width:100%;box-sizing:border-box;">
✍️ <b>پاسخ تشریحی بخش ششم:</b><br>

نتایج برای logistic regression به صورت زیر شدند:

Accuracy: 0.964<br>
Precision: 0.91<br>
Recall: 0.973<br>
F1 Score: 0.94<br>

و همچنین برای naive bayes به صورت زیر شدند:

Accuracy: 0.954<br>
Precision: 0.899<br>
Recall: 0.947<br>
F1 Score: 0.922<br>

مشاهده می کنیم که در تمام معیار ها، logistic regression مقداری بهتر عمل کرده است. در واقع بهتر بودن precision برای آن نشان می دهد که اشتباه کمتری دارد که پیام ham را اسپم تشخیص دهد و در نتیجه برای کاربرد هایی مانند جلوگیری از حذف ایمیل‌ های مهم بهتر است (البته در اینجاچون در هر ۴ معیار بهتر عمل کرده، کلا برتری دارد اما تحلیل مربوط به هر کدام از معیار ها را دارم می گویم). همچنین recall بالاتر هم نشان می دهد که تعداد بیشتری از spam ها را دارد شناسایی می کند.

در واقع naive bayes یک مدل بسیار ساده است که فرض استقلال بین کلمات را دارد. این فرض در داده‌ های واقعی متن تقریبا هیچوقت برقرار نیست اما logistic regression این محدودیت را ندارد و وزن هریک از کلمات را به طور مستقیم یاد می‌گیرد. بنابراین برای پردازش متن با ساختارهای پیچیده‌ تر عملکرد بهتری نشان می‌دهد. البته اینجا عملکرد دو مدل بسیار به هم نزدیک است و naive bayes هم بسیار خوب عمل کرده است و یک برتری آن نسبت به logistic regression این است که بسیار سریع است و کلا زمانی خوب عمل می کند که داده ها زیاد باشند، ویژگی ها sparse باشند (که در اینجا ما bow داریم و این شرط برقرار است) و رابطه بین ویژگی ها ساده باشد و در این مسئله می بینیم که naive bayes هم دارد خوب عمل می کند.

</div>


# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>سوال دوم: استفاده از Logistic Regression و Naive Bayes <br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">هدف این سوال، بررسی و مقایسه عملکرد مدل‌های Logistic Regression و Naive Bayes در تشخیص لینک‌های فیشینگ از لینک‌های قانونی (معتبر) است.
شما باید با استفاده از مجموعه‌داده‌ی ارائه‌شده (آدرس‌های اینترنتی خام)، یک سری ویژگی‌های عددی از این URLها استخراج نمایید و تأثیر انتخاب ویژگی‌ها و نرمال‌سازی داده‌ها را بر عملکرد مدل‌ها بررسی کنید.<br>نکته: در این سوال مجاز به استفاده از کتاب‌خانه‌های آماده هستید.


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش اول: ویژگی‌های ساختاری</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
سه ویژگی زیر را از آدرس استخراج کنید:
<br>
- تعداد نقطه‌ها (nb_dots)
<br>
- تعداد اسلش‌ها (nb_slashes)
<br>
- تعداد خط ‌تیره‌ها (nb_hyphens)
<br>
سپس با این سه ویژگی مدل‌ها را روی هشتاد درصد داده‌ها (داده‌های آموزش) آموزش دهید و روی بیست درصد داده‌ها (داده‌های آزمون) مقایسه کنید
<br>
<br>
- آیا لازم است این ویژگی‌ها را نرمال‌سازی کنید؟ چرا؟ اگر پاسخ شما «بله» است، نوع نرمال‌سازی را توضیح دهید و آن را اجرا کنید.
<br>
- به نظر شما از بین نسخه‌های مختلف Naive Bayes (GaussianNB یا MultinomialNB) کدام برای داده‌های شما مناسب‌تر است؟ دلیل خود را بنویسید و با آن مدل آزمایش را انجام دهید.
</p>

<div dir="rtl" style="background:#fffbe6;font-family: Vazir;border:1px dashed #f0ad4e;padding:12px;border-radius:8px;color:#111;line-height:1.9;overflow:auto;max-width:100%;box-sizing:border-box;">
✍️ <b>پاسخ تشریحی بخش اول:</b><br>

<b>پاسخ سوال اول:</b>

ویژگی هایی که استخراج کردیم شامل تعداد نقطه ها، تعداد اسلش ها و تعداد خط تیره ها در هر url است. این ویژگی ها عددی و گسسته هستند و مقادیر آنها می تواند از صفر تا چندین عدد باشد و در هر url متفاوت باشد. برای logistic regression، نرمال سازی لازم است زیرا نسبت به اندازه ویژگی ها حساس است و اگر یک ویژگی مانند nb_slashes مقادیر بزرگتری از nb_hyphens داشته باشد، باعث می شود وزن مربوط به آن به صورت مصنوعی و غیر واثعی بزرگتر شود و یادگیری مدل هم کند تر می شود. در واقع اگر داده نرمال‌سازی نشود، مقدار ویژگی‌ها در مقیاس‌های مختلف قرار می‌گیرد و سرعت یادگیری کاهش می‌یابد. در واقع به صورت کلی، ویژگی های شمارشی ممکن است توزیع های متفاوتی داشته باشند (مثلا url ها اکثرا ۲ تا ۵ تا slash دارند اما تعداد hyphen ها اغلب کم است) و این ممکن است باعث غلبه کردن ویژگی ها بر هم به صورت مصنوعی شود.

برای GaussianNB فرض می کند که هر ویژگی از توزیع نرمال پیروی می کند و همچنین با مقادیر پیوسته بهتر کار می کند و بنابراین نرمال سازی می تواند مفید باشد. اما برای MultinomialNB، ویژگی ها باید غیر منفی باشند و نرمال سازی ممکن است باعث شود که برخی مقادیر منفی و یا پیوسته شوند که برای این مدل مناسب نیست چرا که این مدل ذاتا با تعداد کار می کند و بنابراین لازم به انجام نرمال سازی در آن نیست و عملکرد مدل را مختل می کند. بنابراین ورودی ای که به MultinomialNB می دهم نرمالایز نشده و ورودی ای که به GaussianNB می دهم نرمالایز شده است و با این فرض مقایسه را انجام می دهم.

برای نوع نرمال سازی هم از Standardization استفاده می کنیم چرا که ویژگی های شمارشی ما در این مسئله کوچک و حدودا در بازه ۰ تا ۵ هستند و تعداد ها معمولا چولگی دارند. کاری که Standardization انجام می دهد این است که میانگین را از هر مقدار کم می کند و سپس بر انحراف معیار تقسیم می کند و داده را حول میانگین متمرکز می‌کند.

<b>پاسخ سوال دوم:</b>

برای داده‌های ما، MultinomialNB مناسب‌تر است. دلیل این انتخاب این است که ویژگی‌های ما شمارشی و غیر منفی هستند (تعداد نقطه‌ها، اسلش‌ها و خط‌تیره‌ها) و MultinomialNB برای داده‌های شمارشی طراحی شده است. در حالی که GaussianNB برای داده ها با ویژگی های پیوسته و با توزیع نرمال استفاده می شود و مقدار ویژگی ها در آن می تواند منفی، مثبت و یا اعشاری باشد.. بنابراین، استفاده از MultinomialNB باعث می‌شود که مدل بتواند الگوهای موجود در داده‌های شمارشی را بهتر یاد بگیرد و عملکرد بهتری داشته باشد. البته با توجه به نتایج گرفته شده، به صورت کلی انگار GaussianNB دارد در این سوال و برای این دیتاست بهتر عمل می کند(accuracy و precision بالاتری دارد) که این بر خلاف تحلیلی است که در بالا ارائه دادم. دلیلش هم می تواند این باشد که سه ویژگی ای که داریم در اینجا بررسی می کنیم،دامنه محدود و کوچکی در حدود ۰ تا ۵ عدد دارند و همین تعداد کم و نزدیک به هم باعث می شود که multinomial احتمالا نتواند ویژگی ها را به خوبی از هم متمایز کند و در واقع چون داده ها کوچک و دارای واریانس کم هستند، gaussian بهتر می تواند آن را مدل کند زیرا بر اساس میانگین و واریانس و نرمال بودن داده ها عمل می کند.

 در واقع در این شرایط، تفاوت‌ های کوچک بین نمونه‌ها بیشتر از طریق میانگین و واریانس قابل مدل‌ سازی است و GaussianNB این الگوهای ظریف را بهتر یاد می‌گیرد.از سوی دیگر، MultinomialNB برای داده‌هایی مناسب است که تعداد رخداد ویژگی‌ها گسترده‌ تر و اطلاعات بیشتری فراهم کنند. اما در این مسئله، ویژگی‌ ها تعداد های کوچک و کم‌ تنوع دارند، بنابراین MultinomialNB در تشخیص کلاس صحیح دچار خطاهای بیشتری می‌شود، هرچند Recall بالاتری دارد و نمونه‌ های فیشینگ را بیشتر شناسایی می‌کند.

</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- جدولی (دیتافریم) شامل ۳ ویژگی استخراج‌شده + برچسب هدف (status)<br>
- جدول مقایسه‌ای شامل Accuracy, Precision, Recall, F1-score برای هر مدل
</p>

In [58]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.naive_bayes import GaussianNB, MultinomialNB
import tldextract

In [59]:
df_1 = pd.read_csv('./datasets/q2/urls.csv')
df_1['nb_dots'] = df_1['url'].str.count(r'\.')
df_1['nb_hyphens'] = df_1['url'].str.count('-')
df_1['nb_slashes'] = df_1['url'].str.count('/')
df_1.head()

url      status  nb_dots  \
0              http://www.crestonwood.com/router.php  legitimate        3   
1  http://shadetreetechnology.com/V4/validation/a...    phishing        1   
2  https://support-appleld.com.secureupdate.duila...    phishing        4   
3                                 http://rgipt.ac.in  legitimate        2   
4  http://www.iracing.com/tracks/gateway-motorspo...  legitimate        2   

   nb_hyphens  nb_slashes  
0           0           3  
1           0           5  
2           1           5  
3           0           2  
4           2           5

In [60]:
def split_dataset(X, y, test_size=0.2, random_state=30):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=test_size, random_state=random_state)
    X_train.reset_index(drop=True, inplace=True)
    X_test.reset_index(drop=True, inplace=True)
    y_train.reset_index(drop=True, inplace=True)
    y_test.reset_index(drop=True, inplace=True)
    return X_train, X_test, y_train, y_test

In [61]:
def evaluate_model(model_name, X_train, X_test, y_test, y_train):
    if model_name == "Logistic Regression":
        model = LogisticRegression()
    elif model_name == "GaussianNB":
        model = GaussianNB()
    elif model_name == "MultinomialNB":
        model = MultinomialNB()
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    print(f"{model_name} Results:")
    print(f"Accuracy: {get_accuracy(y_test, y_pred):.3f}")
    print(f"Precision: {get_precision(y_test, y_pred):.3f}")
    print(f"Recall: {get_recall(y_test, y_pred):.3f}")
    print(f"F1 Score: {f1_score(y_test, y_pred):.3f}")
    print("\n")
    print(classification_report(y_test, y_pred))

In [62]:
X_1 = df_1[["nb_dots", "nb_slashes", "nb_hyphens"]]
y_1 = df_1["status"].map({"legitimate": 0, "phishing": 1})
X_train_1, X_test_1, y_train_1, y_test_1 = split_dataset(X_1, y_1)

In [63]:
X_train_1.head()

nb_dots  nb_slashes  nb_hyphens
0        2           4           0
1        2           4           0
2        2           4           0
3        2           3           0
4        2           2           0

In [64]:
X_train_1.shape, X_test_1.shape

((9144, 3), (2286, 3))

In [65]:
y_train_1.shape, y_test_1.shape

((9144,), (2286,))

In [66]:
scaler_1 = StandardScaler()
X_train_scaled_1 = scaler_1.fit_transform(X_train_1)
X_test_scaled_1 = scaler_1.transform(X_test_1)

In [67]:
X_train_scaled_1

array([[-0.3570172 , -0.15216637, -0.47500846],
       [-0.3570172 , -0.15216637, -0.47500846],
       [-0.3570172 , -0.15216637, -0.47500846],
       ...,
       [-0.3570172 ,  0.90755129,  6.31773993],
       [-1.10783718, -0.15216637,  0.01018785],
       [-0.3570172 , -0.6820252 , -0.47500846]], shape=(9144, 3))

In [47]:
X_train_1.head()

nb_dots  nb_slashes  nb_hyphens
0        2           4           0
1        2           4           0
2        2           4           0
3        2           3           0
4        2           2           0

In [48]:
evaluate_model("Logistic Regression", X_train_1, X_test_1, y_test_1, y_train_1)

Logistic Regression Results:
Accuracy: 0.672
Precision: 0.719
Recall: 0.587
F1 Score: 0.647


              precision    recall  f1-score   support

           0       0.64      0.76      0.69      1120
           1       0.72      0.59      0.65      1166

    accuracy                           0.67      2286
   macro avg       0.68      0.67      0.67      2286
weighted avg       0.68      0.67      0.67      2286



In [ ]:
# without normalization
evaluate_model("GaussianNB", X_train_scaled_1, X_test_scaled_1, y_test_1, y_train_1)

GaussianNB Results:
Accuracy: 0.666
Precision: 0.783
Recall: 0.479
F1 Score: 0.594


              precision    recall  f1-score   support

           0       0.61      0.86      0.72      1120
           1       0.78      0.48      0.59      1166

    accuracy                           0.67      2286
   macro avg       0.70      0.67      0.66      2286
weighted avg       0.70      0.67      0.65      2286



In [ ]:
# with normalization
evaluate_model("GaussianNB", X_train_1, X_test_1, y_test_1, y_train_1)

GaussianNB Results:
Accuracy: 0.666
Precision: 0.783
Recall: 0.479
F1 Score: 0.594


              precision    recall  f1-score   support

           0       0.61      0.86      0.72      1120
           1       0.78      0.48      0.59      1166

    accuracy                           0.67      2286
   macro avg       0.70      0.67      0.66      2286
weighted avg       0.70      0.67      0.65      2286



In [11]:
evaluate_model("MultinomialNB", X_train_1, X_test_1, y_test_1, y_train_1)

MultinomialNB Results:
Accuracy: 0.560
Precision: 0.550
Recall: 0.750
F1 Score: 0.635


              precision    recall  f1-score   support

           0       0.58      0.36      0.45      1120
           1       0.55      0.75      0.63      1166

    accuracy                           0.56      2286
   macro avg       0.57      0.56      0.54      2286
weighted avg       0.57      0.56      0.54      2286



In [ ]:
# MultinomialNB will not work with negative values (and thus doesn't need normalization)
evaluate_model("MultinomialNB", X_train_scaled_1, X_test_scaled_1, y_test_1, y_train_1)

ValueError: Negative values in data passed to MultinomialNB (input X).

## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش دوم: ویژگی‌های آماری و محتوایی</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
سه ویژگی زیر را از آدرس استخراج کنید:
<br>
- طول کل آدرس (length_url)
<br>
- نسبت تعداد ارقام (0–9) به کل طول آدرس (ratio_digits_url)
<br>
- طولانی‌ترین کلمه (کاراکترهای الفبایی) در URL که با علائم جداساز (نقطه، علامت‌سوال، اسلش و...) از هم جدا شده‌اند. (longest_words_raw)
<br>
سپس با این سه ویژگی مدل‌ها را روی هشتاد درصد داده‌ها (داده‌های آموزش) آموزش دهید و روی داده‌های آزمون (بیست درصد داده‌ها) مقایسه کنید.
<br>
- به نظر شما استفاده از Naive Bayes برای این ویژگی‌ها منطقی است؟ توضیح دهید. کدام نسخه مناسب‌تر است؟ 
</p>

<div dir="rtl" style="background:#fffbe6;font-family: Vazir;border:1px dashed #f0ad4e;padding:12px;border-radius:8px;color:#111;line-height:1.9;overflow:auto;max-width:100%;box-sizing:border-box;">
✍️ <b>پاسخ تشریحی بخش دوم:</b><br>
در اینجا می توان از naive bayes استفاده کرد اما نوع آن مهم است.

سه ویژگی استخراج‌شده از URLها شامل طول کل آدرس، نسبت تعداد ارقام به طول آدرس و طولانی‌ترین کلمه‌ی موجود در URL هستند. این سه ویژگی همگی ویژگی‌ های پیوسته یا غیر شمارشی محسوب می‌ شوند. در واقع ویژگی طول کل آدرس یک ویژگی است که از نوع طول است و می تواند بسیار متغیر باشد. همچنین نسبت تعداد ارقام به کل طول آدرس هم یک ویژگی پیوسته است و در نهایت طولانی ترین کلمه هم باز می تواند طول متغیری داشته باشد (که الته متغیر بودن آن از طول کل آدرس کمتر است و احتمالا طول آن حداکثر ۲۰ یا ۳۰ باشد).  و هیچ‌ کدام از نوع شمارشی (count features) نیستند. از آن‌جایی که مدل Multinomial Naive Bayes برای داده‌های شمارشی طراحی شده است، استفاده از آن برای این نوع ویژگی‌ها منطقی و مناسب نیست.

در مقابل، GaussianNB  فرض می‌کند که هر ویژگی دارای یک توزیع نرمال است و برای ویژگی‌های پیوسته مناسب‌تر است. به همین دلیل، در بین نسخه‌های Naive Bayes تنها GaussianNB برای این سه ویژگی قابل استفاده و منطقی است.

بنابراین آزمایش‌ ها برای این بخش با مدل GaussianNB انجام شد (البته با multinomialNB هم انجام دادم و خروجی گرفتم که نتیجه آن بسیار بد است و به صورت منطقی هم به دلیل وجود ویژگی پیوسته قابل استفاده نیست).

</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- جدولی (دیتافریم) شامل ۳ ویژگی استخراج‌شده + برچسب هدف (status)<br>
- جدول مقایسه‌ای شامل Accuracy, Precision, Recall, F1-score برای هر مدل
</p>

In [69]:
df_2 = pd.read_csv('./datasets/q2/urls.csv')
df_2['length_url'] = df_2['url'].str.len()
df_2['count_digits'] = df_2['url'].str.count(r'\d')
df_2['ratio_digits_url'] = df_2['count_digits'] / df_2['length_url']
all_words = df_2['url'].str.findall(r'[a-zA-Z]+')
df_2['longest_words_raw'] = all_words.apply(lambda words_list: len(max(words_list, key=len)))
df_2 = df_2.drop(columns=['count_digits'])

In [70]:
df_2.head()

url      status  length_url  \
0              http://www.crestonwood.com/router.php  legitimate          37   
1  http://shadetreetechnology.com/V4/validation/a...    phishing          77   
2  https://support-appleld.com.secureupdate.duila...    phishing         126   
3                                 http://rgipt.ac.in  legitimate          18   
4  http://www.iracing.com/tracks/gateway-motorspo...  legitimate          55   

   ratio_digits_url  longest_words_raw  
0          0.000000                 11  
1          0.220779                 19  
2          0.150794                 13  
3          0.000000                  5  
4          0.000000                 11

In [52]:
X_2 = df_2[["length_url", "ratio_digits_url", "longest_words_raw"]]
y_2 = df_2["status"].map({"legitimate": 0, "phishing": 1})
X_train_2, X_test_2, y_train_2, y_test_2 = split_dataset(X_2, y_2)

In [53]:
X_train_2.head()

length_url  ratio_digits_url  longest_words_raw
0          41          0.121951                  9
1          39          0.000000                  9
2          47          0.000000                  9
3          33          0.000000                 11
4          29          0.000000                 15

In [54]:
scaler_2 = StandardScaler()
X_train_scaled_2 = scaler_2.fit_transform(X_train_2)
X_test_scaled_2 = scaler_2.transform(X_test_2)

In [56]:
evaluate_model("Logistic Regression", X_train_scaled_2, X_test_scaled_2, y_test_2, y_train_2)

Logistic Regression Results:
Accuracy: 0.659
Precision: 0.730
Recall: 0.526
F1 Score: 0.611


              precision    recall  f1-score   support

           0       0.62      0.80      0.70      1120
           1       0.73      0.53      0.61      1166

    accuracy                           0.66      2286
   macro avg       0.67      0.66      0.65      2286
weighted avg       0.67      0.66      0.65      2286



In [521]:
evaluate_model("GaussianNB", X_train_scaled_2, X_test_scaled_2, y_test_2, y_train_2)

GaussianNB Results:
Accuracy: 0.634
Precision: 0.780
Recall: 0.395
F1 Score: 0.524


              precision    recall  f1-score   support

           0       0.58      0.88      0.70      1120
           1       0.78      0.39      0.52      1166

    accuracy                           0.63      2286
   macro avg       0.68      0.64      0.61      2286
weighted avg       0.68      0.63      0.61      2286



In [522]:
evaluate_model("MultinomialNB", X_train_2, X_test_2, y_test_2, y_train_2)

MultinomialNB Results:
Accuracy: 0.567
Precision: 0.592
Recall: 0.491
F1 Score: 0.536


              precision    recall  f1-score   support

           0       0.55      0.65      0.59      1120
           1       0.59      0.49      0.54      1166

    accuracy                           0.57      2286
   macro avg       0.57      0.57      0.57      2286
weighted avg       0.57      0.57      0.56      2286



## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش سوم: ویژگی‌های خلاقانه</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
سه ویژگی به انتخاب خودتان از آدرس‌ها استخراج نمایید و مشابه دو بخش قبلی، مدل‌ها را آموزش دهید.
برای هر ویژگی توضیح دهید چرا ممکن است در تشخیص فیشینگ مؤثر باشد؟
</p>

<div dir="rtl" style="background:#fffbe6;font-family: Vazir;border:1px dashed #f0ad4e;padding:12px;border-radius:8px;color:#111;line-height:1.9;overflow:auto;max-width:100%;box-sizing:border-box;">
✍️ <b>پاسخ تشریحی بخش سوم:</b><br>

سه ویژگی انتحابی من به صورت زیر هستند:

<b>۱. تعداد subdomain ها</b>

URLهای phishing معمولا از ساب‌دامین‌های طولانی و زیاد استفاده می‌کنند و با زیاد کردن ساب دامین ها قصد دارند صرفا پیچیدگی را افزایش دهند. برای URL های معمولی، این تعداد در حدود ۱ تا ۳ تا است.


<b>۲. طولانی ترین رشته ای که هم شامل عدد باشد و هم حرف </b>

ویژگی ای که در قسمت قبل معرفی شده بود، فقط برای طولانی ترین رشته شامل کاراکتر های الفبایی بین علائم جداساز بود. اما در این ویژگی، طولانی ترین رشته ای که هم شامل عدد باشد و هم حرف را استخراج می کنیم. دلیل اهمیت این ویژگی این است که سایت های فیشینگ معمولا از ترکیب حروف و اعداد در URL استفاده می کنند تا شبیه به سایت های قانونی به نظر برسند و این ترکیب می تواند یک نشانه مهم برای تشخیص phishing باشد.

<b>۳. تعداد علائم نگارشی و کاراکتر های مشکوک زیاد </b>

URLهای phishing معمولا شامل تعداد زیادی کاراکتر رمزگذاری‌ شده یا ویژه هستند‌(مانند %, ?, =, &, _, !, $, *, ;).استفاده زیاد از این کاراکتر ها معمولا برای پنهان کردن URL واقعی و ایجاد پیچیدگی مصنوعی و غیر طبیعی است و می تواند از نشانه های phishing باشد. در واقع URL های قانونی معمولا ساختار ساده ای دارند.
</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- جدولی (دیتافریم) شامل ۳ ویژگی استخراج‌شده + برچسب هدف (status)<br>
- جدول مقایسه‌ای شامل Accuracy, Precision, Recall, F1-score برای هر مدل
</p>

In [71]:
df_3 = pd.read_csv('./datasets/q2/urls.csv')
all_words_w_nums = df_3['url'].str.findall(r'[a-zA-Z]+')
df_3['longest_strings'] = all_words_w_nums.apply(lambda words_list: len(max(words_list, key=len)))
punc_pattern = r"[%;&=\?\$!\*_\+~]"
df_3["punctuations_num"] = df_3["url"].str.count(punc_pattern)

def count_subdomains(url):
    ext = tldextract.extract(url)
    if ext.subdomain == "":
        return 0
    return len(ext.subdomain.split("."))

df_3["subdomains_num"] = df_3["url"].apply(count_subdomains)

In [72]:
df_3.head()

url      status  \
0              http://www.crestonwood.com/router.php  legitimate   
1  http://shadetreetechnology.com/V4/validation/a...    phishing   
2  https://support-appleld.com.secureupdate.duila...    phishing   
3                                 http://rgipt.ac.in  legitimate   
4  http://www.iracing.com/tracks/gateway-motorspo...  legitimate   

   longest_strings  punctuations_num  subdomains_num  
0               11                 0               1  
1               19                 0               0  
2               13                 8               3  
3                5                 0               0  
4               11                 0               1

In [527]:
X_3 = df_3[["longest_strings", "punctuations_num", "subdomains_num"]]
y_3 = df_3["status"].map({"legitimate": 0, "phishing": 1})
X_train_3, X_test_3, y_train_3, y_test_3 = split_dataset(X_3, y_3)

In [528]:
X_train_3.head()

longest_strings  punctuations_num  subdomains_num
0                9                 0               1
1                9                 0               1
2                9                 1               1
3               11                 0               1
4               15                 0               1

In [529]:
scaler_3 = StandardScaler()
X_train_scaled_3 = scaler_3.fit_transform(X_train_3)
X_test_scaled_3 = scaler_3.transform(X_test_3)

In [530]:
evaluate_model("Logistic Regression", X_train_scaled_3, X_test_scaled_3, y_test_3, y_train_3)

Logistic Regression Results:
Accuracy: 0.592
Precision: 0.649
Recall: 0.436
F1 Score: 0.521


              precision    recall  f1-score   support

           0       0.56      0.75      0.64      1120
           1       0.65      0.44      0.52      1166

    accuracy                           0.59      2286
   macro avg       0.61      0.60      0.58      2286
weighted avg       0.61      0.59      0.58      2286



In [531]:
evaluate_model("GaussianNB", X_train_scaled_3, X_test_scaled_3, y_test_3, y_train_3)

GaussianNB Results:
Accuracy: 0.691
Precision: 0.768
Recall: 0.565
F1 Score: 0.651


              precision    recall  f1-score   support

           0       0.64      0.82      0.72      1120
           1       0.77      0.57      0.65      1166

    accuracy                           0.69      2286
   macro avg       0.71      0.69      0.69      2286
weighted avg       0.71      0.69      0.69      2286



In [532]:
evaluate_model("MultinomialNB", X_train_3, X_test_3, y_test_3, y_train_3)

MultinomialNB Results:
Accuracy: 0.558
Precision: 0.658
Recall: 0.276
F1 Score: 0.389


              precision    recall  f1-score   support

           0       0.53      0.85      0.65      1120
           1       0.66      0.28      0.39      1166

    accuracy                           0.56      2286
   macro avg       0.59      0.56      0.52      2286
weighted avg       0.60      0.56      0.52      2286



## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش چهارم: ترکیب همه ویژگی‌ها</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
تمام ۹ ویژگی (۳ ساختاری + ۳ آماری + ۳ خلاقانه) را با هم ترکیب کنید.
سپس هر دو مدل را دوباره آموزش دهید و نتایج را ارزیابی کنید.
<br>
آیا ترکیب همه ویژگی‌ها باعث بهبود عملکرد مدل می‌شود؟
<br>
اگر خیر، به نظر شما دلیل آن چیست؟ (مثلاً تداخل بین ویژگی‌ها یا همبستگی بالا و...)
</p>


<div dir="rtl" style="background:#fffbe6;font-family: Vazir;border:1px dashed #f0ad4e;padding:12px;border-radius:8px;color:#111;line-height:1.9;overflow:auto;max-width:100%;box-sizing:border-box;">
✍️ <b>پاسخ تشریحی بخش چهارم:</b><br>

در اینجا در ابتدا بگویم که به دلیل وجود ویژگی های پیوسته و غیر شمارشی زیاد بین این ۹ ویژگی (مطابق قسمت های قبل که بررسی شد)، GaussianNB به وضوح انتخاب بهتر و درست برای naive bayes است و MultinomialNB را صرفا نتایجش را قرار دادم تا متوجه شویم که برای این ویژگی ها اصلا مناسب نیست و خوب عمل نمی کند.

نتایج نشان می دهد که عملکرد مدل‌ها نسبت به زمانی که تنها ۳ ویژگی استفاده شده بود بهتر شد (البته نتایج مربوط به Multinomial بدتر شد که طبق توضیحاتی که دادم و نتایجی که از قسمت های قبل بدست آمد، مدل انتخابی ما نیست و Gaussian را انتخاب کردم و نتایج مربوط به Multinomial را صرفا قرار دادم که موجود باشد). دلیل اصلی این بهبود آن است که هر مجموعه از ویژگی‌ ها بخشی از خصوصیات URL را پوشش می‌دهد و ترکیب آن‌ها باعث می‌شود مدل از چندین زاویه مختلف به الگوهای phishing دسترسی داشته باشد. ویژگی‌های ساختاری الگوهای ظاهری آدرس را به مدل می‌دهند، ویژگی‌های آماری پیچیدگی و مشخصات طولی URL را بیان می‌کنند و ویژگی‌ های خلاقانه ای که تعریف کردم بیشتر و به صورت عمیق تر به استخراج ویژگی های مربوط به phising نزدیک می شود و آن را با دیدن نمونه های واقعی از URL های phishing تعریف کردم که باعث می شود به صورت مستقیم تر بتواند درباره آنها بر اساس این ۳ ویژگی تصمیم گیری کند و ۶ ویژگی دیگر می توانند مکمل خوبی برای این ۳ ویژگی باشند. استفاده همزمان از این سه نوع ویژگی باعث می‌شود مدل تصمیم‌ گیری کامل‌تری انجام دهد و خطاهای مربوط به تکیه بر یک نوع ویژگی کاهش یابد.

همچنین با وجود اینکه برخی از ۹ ویژگی ممکن است همبستگی یا تداخل داشته باشند، نتایج نشان داد که عملکرد هر دو مدل Logistic Regression و GaussianNB پس از ترکیب ویژگی‌ها بهتر شده است. دلیل این موضوع آن است که ماهیت ویژگی‌ ها متفاوت است و هر گروه از ویژگی‌ ها بخشی جداگانه از رفتار URL را توصیف می‌کند. ویژگی‌ های ساختاری، آماری و خلاقانه هر کدام اطلاعات متفاوت اما مکمل ارائه می‌دهند و در نتیجه مدل با دید کامل‌تری نسبت به ساختار و رفتار URLها تصمیم‌گیری می‌کند. التبه بعضی از ویژگی ها مانند تعداد نقاط و تعداد subdomain ها با هم همبستگی دارند اما باز هم به دلیل وجود ویژگی های متنوع دیگر، تاثیر این همبستگی کمتر از تاثیر تنوع ویژگی ها می شود.

علاوه بر این، Logistic Regression با استفاده از Regularization (که در کتابخانه به صورت پیش فرص استفاده می شود)نسبت به همبستگی بین ویژگی‌ ها مقاوم است و وزن‌دهی مناسب به ویژگی‌ های مهم را یاد می‌گیرد. GaussianNB نیز به دلیل سادگی و مدل‌سازی مستقل هر ویژگی، از تداخل ویژگی‌ ها آسیب جدی نمی‌بیند و حتی از نشانه‌های تکراری برای تقویت تصمیم خود استفاده می‌کند.
در نتیجه، ترکیب ویژگی‌ها نه تنها باعث تداخل نشده، بلکه به دلیل مکمل بودن آن‌ها، عملکرد مدل‌ها را بهبود داده است.
</div>


In [73]:
df_merged = df_1.merge(df_2, on=["url", "status"]).merge(df_3, on=["url", "status"])
df_merged.head()


url      status  nb_dots  \
0              http://www.crestonwood.com/router.php  legitimate        3   
1  http://shadetreetechnology.com/V4/validation/a...    phishing        1   
2  https://support-appleld.com.secureupdate.duila...    phishing        4   
3                                 http://rgipt.ac.in  legitimate        2   
4  http://www.iracing.com/tracks/gateway-motorspo...  legitimate        2   

   nb_hyphens  nb_slashes  length_url  ratio_digits_url  longest_words_raw  \
0           0           3          37          0.000000                 11   
1           0           5          77          0.220779                 19   
2           1           5         126          0.150794                 13   
3           0           2          18          0.000000                  5   
4           2           5          55          0.000000                 11   

   longest_strings  punctuations_num  subdomains_num  
0               11                 0               1  
1               19                 0               0  
2               13                 8               3  
3                5                 0               0  
4               11                 0               1

In [74]:
X_merged = df_merged.drop(columns=['url', 'status'])
y_merged = df_merged["status"].map({"legitimate": 0, "phishing": 1})
X_train_merged, X_test_merged, y_train_merged, y_test_merged = split_dataset(X_merged, y_merged)

In [75]:
X_train_merged.head()

nb_dots  nb_hyphens  nb_slashes  length_url  ratio_digits_url  \
0        2           1           3          29          0.000000   
1        2           0           4          39          0.000000   
2        4           4           3          83          0.048193   
3        2           0           4          57          0.000000   
4        2           0           2          29          0.000000   

   longest_words_raw  longest_strings  punctuations_num  subdomains_num  
0                  8                8                 0               1  
1                  9                9                 0               1  
2                  6                6                 6               1  
3                 22               22                 1               0  
4                 15               15                 0               1

In [76]:
scaler_merged = StandardScaler()
X_train_scaled_merged = scaler_merged.fit_transform(X_train_merged)
X_test_scaled_merged = scaler_merged.transform(X_test_merged)

In [77]:
X_train_scaled_merged

array([[-3.60368398e-01,  6.75012353e-04, -6.97080996e-01, ...,
        -6.86670247e-01, -3.18077473e-01,  4.22686740e-02],
       [-3.60368398e-01, -4.74325988e-01, -1.61926838e-01, ...,
        -4.86496275e-01, -3.18077473e-01,  4.22686740e-02],
       [ 1.11530309e+00,  1.42567801e+00, -6.97080996e-01, ...,
        -1.08701819e+00,  1.37381480e+00,  4.22686740e-02],
       ...,
       [-3.60368398e-01,  6.17568802e+00,  9.08381479e-01, ...,
        -4.86496275e-01, -3.18077473e-01,  4.22686740e-02],
       [-1.09820414e+00,  6.75012353e-04, -1.61926838e-01, ...,
        -1.08701819e+00, -3.18077473e-01, -1.10513141e+00],
       [-3.60368398e-01, -4.74325988e-01, -6.97080996e-01, ...,
        -8.61483307e-02, -3.18077473e-01,  4.22686740e-02]],
      shape=(9148, 9))

In [ ]:
# with normalization
evaluate_model("Logistic Regression", X_train_scaled_merged, X_test_scaled_merged, y_test_merged, y_train_merged)

Logistic Regression Results:
Accuracy: 0.727
Precision: 0.782
Recall: 0.630
F1 Score: 0.698


              precision    recall  f1-score   support

           0       0.69      0.82      0.75      1142
           1       0.78      0.63      0.70      1146

    accuracy                           0.73      2288
   macro avg       0.74      0.73      0.72      2288
weighted avg       0.74      0.73      0.72      2288



In [ ]:
# without normalization
evaluate_model("Logistic Regression", X_train_merged, X_test_merged, y_test_merged, y_train_merged)

Logistic Regression Results:
Accuracy: 0.687
Precision: 0.743
Recall: 0.573
F1 Score: 0.647


              precision    recall  f1-score   support

           0       0.65      0.80      0.72      1142
           1       0.74      0.57      0.65      1146

    accuracy                           0.69      2288
   macro avg       0.70      0.69      0.68      2288
weighted avg       0.70      0.69      0.68      2288



/home/amin/.local/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [ ]:
# without normalization
evaluate_model("GaussianNB", X_train_merged, X_test_merged, y_test_merged, y_train_merged)

GaussianNB Results:
Accuracy: 0.711
Precision: 0.867
Recall: 0.500
F1 Score: 0.634


              precision    recall  f1-score   support

           0       0.65      0.92      0.76      1142
           1       0.87      0.50      0.63      1146

    accuracy                           0.71      2288
   macro avg       0.76      0.71      0.70      2288
weighted avg       0.76      0.71      0.70      2288



In [82]:
# with normalization
evaluate_model("GaussianNB", X_train_scaled_merged, X_test_scaled_merged, y_test_merged, y_train_merged)

GaussianNB Results:
Accuracy: 0.711
Precision: 0.867
Recall: 0.500
F1 Score: 0.634


              precision    recall  f1-score   support

           0       0.65      0.92      0.76      1142
           1       0.87      0.50      0.63      1146

    accuracy                           0.71      2288
   macro avg       0.76      0.71      0.70      2288
weighted avg       0.76      0.71      0.70      2288



In [540]:
evaluate_model("MultinomialNB", X_train_merged, X_test_merged, y_test_merged, y_train_merged)

MultinomialNB Results:
Accuracy: 0.614
Precision: 0.710
Recall: 0.387
F1 Score: 0.501


              precision    recall  f1-score   support

           0       0.58      0.84      0.68      1142
           1       0.71      0.39      0.50      1146

    accuracy                           0.61      2288
   macro avg       0.64      0.61      0.59      2288
weighted avg       0.64      0.61      0.59      2288



## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش پنجم: جمع‌بندی نهایی</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
نتیجه‌گیری نهایی خودتان از این سوال را ارائه دهید. برای مثال باید "حداقل" به سوال‌های زیر پاسخ دهید:
<br>
- کدام مدل برای این نوع داده بهتر عمل کرده است؟ چرا؟
<br>
- کدام نوع ویژگی بیشترین تأثیر را داشته است؟ کدام نوع کمترین؟ چرا؟
<br>
- نرمال‌سازی نیاز بوده است؟ اگر بله، برای هر دو مدل؟ چرا؟
<br>
- غیر از استخراح ویژگی از خود آدرس اینترنتی، چه رویکرد دیگری می‌توان اتخاذ نمود؟
</p>


### نتایج بخش اول (ویژگی‌های ساختاری)
| Model             | Accuracy | Precision | Recall | F1 Score |
|-------------------|----------|-----------|--------|----------|
| Logistic Regression | 0.672    | 0.719     | 0.587  | 0.647    |
| GaussianNB          | 0.666    | 0.783     | 0.479  | 0.594    |
| MultinomialNB       | 0.560    | 0.550     | 0.750  | 0.635    |

### نتایج بخش دوم (ویژگی‌های آماری و محتوایی)
| Model             | Accuracy | Precision | Recall | F1 Score |
|-------------------|----------|-----------|--------|----------|
| Logistic Regression | 0.659    | 0.730     | 0.526  | 0.611    |
| GaussianNB          | 0.634    | 0.780     | 0.395  | 0.524    |
| MultinomialNB       | 0.567    | 0.592     | 0.491  | 0.536    |

### نتایج بخش سوم (ویژگی‌های خلاقانه)
| Model             | Accuracy | Precision | Recall | F1 Score |
|-------------------|----------|-----------|--------|----------|
| Logistic Regression | 0.592    | 0.649     | 0.436  | 0.521    |
| GaussianNB          | 0.691    | 0.768     | 0.565  | 0.651    |
| MultinomialNB       | 0.558    | 0.658     | 0.276  | 0.389    |

### نتایج بخش چهارم (ترکیب همه ۹ ویژگی)
| Model             | Accuracy | Precision | Recall | F1 Score |
|-------------------|----------|-----------|--------|----------|
| Logistic Regression | 0.687    | 0.743     | 0.573  | 0.647    |
| GaussianNB          | 0.711    | 0.867     | 0.500  | 0.634    |
| MultinomialNB       | 0.614    | 0.710     | 0.387  | 0.501    |


<div dir="rtl" style="background:#fffbe6;font-family: Vazir;border:1px dashed #f0ad4e;padding:12px;border-radius:8px;color:#111;line-height:1.9;overflow:auto;max-width:100%;box-sizing:border-box;">
✍️ <b>پاسخ تشریحی بخش پنجم:</b><br>

با مقایسه متریک ها در بخش های قبل برای logistic و gaussian می بینیم که logistic به صورت کلی بهتر عمل کرده است (به خصوص در accuracy و f1).
logistic برخلاف gaussian که یک مدل ساده و مبتنی بر فرضیات آماری است، سعی می‌کند مرز جداکننده بین لینک‌های واقعی و phishing را با یاد گرفتن رابطه مستقیم بین ویژگی‌ ها پیدا کند. logistic بهتر می‌فهمد که ترکیب چند ویژگی مختلف چه معنایی دارد. برای مثال اگر یک URL هم subdomain های زیادی داشته باشد، هم طولانی باشد و هم کاراکتر های عجیب داشته باشد، logistic می‌تواند این چند نشانه را کنار هم بگذارد و به نتیجه برسد.

از طرف دیگر، logistic وقتی داده‌ها را نرمال‌ سازی می‌کنیم، پایدارتر و دقیق‌تر می‌ شوددر واقع همانطور که در قسمت ترکیب ویژگی ها نتایج حاصل از با نرمال سازی و بدون نرمال سازی را برای آن قرار داده ام، مشاهده می شود که warning داده شده و می گوید که در این تعداد iteration نتوانسته converge کند که به این دلیل است که ویژگی ها مقیاسشان بسیار متفاوت است و طول URL مثلا می تواند تا ۱۰۰ هم برود اما تعداد نقطه ها عدد کوچکی است و همچنین نسبت ارقام هم یک عدد کوچک اعشاری است و این اختلاف مقیاس باعث می‌شود در logistic بعضی ویژگی‌ها ده‌ ها برابر بیشتر از بقیه روی تابع هزینه اثر بگذارند. در نتیجه الگوریتم بهینه‌ سازی نمی‌تواند به راحتی حداقل تابع هزینه را پیدا کند، در حلقه به‌ روزرسانی گیر می‌کند و به حد همگرایی نمی‌رسد. نرمال‌ سازی باعث می‌ شود مدل با همه ویژگی‌ها به یک اندازه رفتار کند و وزن‌ دهی درستی یاد بگیرد و این مشکلات به وجود نیاید. اما این برای gaussian متفاوت است. gaussian بر اساس میانگین و واریانس هر ویژگی برای هر کلاس کار می‌کند و فرمول احتمال آن فقط به فاصله نسبی مقدار ویژگی تا میانگین کلاس وابسته است. نرمال‌سازی ( StandardScaler) همین میانگین و واریانس را تغییر می‌دهد، اما برای هر دو کلاس دقیقا به یک شکل انجام می‌ شود و در نتیجه رابطه‌ی بین دو کلاس تغییری نمی‌کند. در واقع gaussian به شکل توزیع فیچر ها حساس است نه به مقیاس آنها. به همین دلیل وقتی داده را نرمال‌سازی می‌کنیم، هر دو کلاس به همان نسبت تغییر می‌کنند و مدل عملا همان تصمیم قبلی را می‌ گیرد. بنابراین نتایج با نرمال‌ سازی و بدون نرمال‌ سازی یکسان باقی می‌مانند.

ویژگی‌های آماری و محتوایی بیشترین نقش را در تشخیص لینک‌ های phishing بازی کرده‌ اند(بر اساس f1 و کلا متریک های دیگر). دلیلش هم این است که این نوع ویژگی‌ ها اطلاعات بسیار واقعی‌ تر و معنادار تری درباره ماهیت یک URL به ما می‌دهند. برای مثال طول خود URL، نسبت عدد به کل متن و طولانی‌ترین کلمه‌های داخل آن معمولا نشانه‌ های واضحی از رفتار phishing هستند. لینک‌های phishing معمولاً طولانی‌تر، شلوغ‌تر و نامنظم‌تر از URLهای معمولی‌اند، بنابراین این ویژگی‌ها تفاوت واقعی بین لینک سالم و لینک phishing را بهتر ثبت می‌کنند. به همین خاطر، وقتی مدل با این ویژگی‌ ها آموزش داده شد، عملکرد آن به‌طور محسوسی بهتر شد.

در مقابل، ویژگی‌های ساختاری ساده مثل تعداد نقطه‌ها، اسلش‌ ها یا خط‌تیره‌ها کمترین تاثیر را داشتند. دلیلش این است که این ویژگی‌ها خیلی کلی و محدود هستند و لزوما رفتار مشخصی برای تفکیک لینک‌های سالم از phishing ندارند. بسیاری از سایت‌های عادی هم ممکن است چند اسلش یا چند نقطه داشته باشند و این ویژگی‌ها همیشه نشانه‌ی phishing نیستند. بنابراین وقتی مدل فقط با این سه ویژگی ساده آموزش داده می‌شود، اطلاعات کافی برای تشخیص دقیق در اختیار ندارد.

همچنین چند رویکرد برای استخراج ویژگی وجود دارد.تشخیص لینک‌های یphishing فقط با بررسی خود URL انجام نمی‌شود و در دنیای واقعی روش‌های قوی‌تر و دقیق‌تری هم وجود دارد. یکی از رویکردهای مهم، استفاده از اطلاعات خارج از ساختار URL است. برای مثال می‌توان از اطلاعات مربوط به دامنه و سرور استفاده کرد. بسیاری از لینک‌های یphishing روی دامنه‌هایی میزبانی می‌شوند که تازه ثبت شده‌اند، مدت مالکیت آن‌ها کوتاه است، یا روی سرورهایی با سابقه عجیب قرار دارند. بنابراین ویژگی‌هایی مثل سن دامنه، اعتبار SSL، IP سرور یا موقعیت جغرافیایی آن می‌توانند نشانه‌های خیلی مهمی باشند.

رویکرد دیگر استفاده از روش‌های مدل‌سازی مبتنی بر متن (NLP) است. به‌جای اینکه خودمان دستی چند ویژگی بسازیم، مدل می‌تواند کل URL را مانند یک جمله یا رشته متنی پردازش کند. روش‌هایی مثل TF-IDF، شمارش n-gramها یا تبدیل کاراکترهای URL به بردارهای معنایی می‌توانند اطلاعات بسیار جزئی و ظریف‌تری استخراج کنند که با چشم انسان یا ویژگی‌سازی دستی قابل تشخیص نیست. این روش‌ها خطاهای مدل را کم می‌کنند و دقت را بالا می‌برند.

یک روش مهم دیگر، تحلیل محتوا و رفتار صفحه مقصد است. یعنی فقط خود لینک را بررسی نکنیم، بلکه ببینیم این لینک بعد از باز شدن چه رفتاری دارد. صفحه‌های یphishing معمولا ظاهر ساده‌تری دارند، فرم‌های ورود جعلی دارند، یا از اسکریپت‌های پنهان برای جمع‌آوری اطلاعات استفاده می‌کنند. در نتیجه، تحلیل HTML، جاوااسکریپت و ساختار صفحه می‌تواند هشدارهای خیلی دقیق‌تر تولید کند.

در نهایت، می‌توان از مدل‌های پیشرفته‌تر یادگیری ماشین یا حتی شبکه‌های عصبی استفاده کرد. مدل‌هایی مثل Random Forest، XGBoost یا CNNهای مبتنی بر کاراکتر می‌توانند روابط پیچیده و الگوهای غیرخطی را بهتر از مدل‌های ساده‌ای مثل Logistic یا Naive Bayes یاد بگیرند. این روش‌ها به‌خصوص زمانی عالی عمل می‌کنند که داده زیاد باشد.
</div>


https://chatgpt.com/share/691612c9-9a80-8003-9c32-a52d011e9a30

# <h1 style="text-align: right;">**نکات مهم و قوانین تحویل**</h1>



<h4 dir="rtl" style="font-family: Vazir; width: 85%;">فایل ارسالی شما باید با فرمت زیر نامگذاری شود: <code>NLP_CA{n}_{LASTNAME}_{STUDENTID}.ipynb</code></h4>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">نحوه انجام تمرین:</h4>
<ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
  <li>سلول‌های کد با برچسب <code>WRITE YOUR CODE HERE</code> را تکمیل کنید.</li>
  <li>برای پاسخ‌های متنی، متن <code>{{پاسخ_خود_را_اینجا_بنویسید}}</code> را با پاسخ خود جایگزین کنید.</li>
</ul>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">صداقت علمی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;"> <li>ما نوت‌بوک‌های تعداد مشخصی از دانشجویان که به صورت تصادفی انتخاب می‌شوند، بررسی خواهیم کرد. این بررسی‌ها اطمینان حاصل می‌کنند که کدی که نوشتید واقعاً پاسخ‌های موجود در نوت‌بوک شما را تولید می‌کند. اگر پاسخ‌های صحیح را در نوت‌بوک خود بدون کدی که واقعاً آن پاسخ‌ها را تولید کند تحویل دهید، این یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> <li>ما همچنین بررسی‌های خودکاری را برای تشخیص سرقت علمی در نوت‌بوک‌های کولب انجام خواهیم داد. کپی کردن کد از دیگران نیز یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> </ul>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">توضیحات تکمیلی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
<li>
خوانایی و دقت بررسی‌ها در گزارش نهایی از اهمیت ویژه‌ای برخوردار است. به تمرین‌هایی که به صورت کاغذی تحویل داده شوند یا به صورت عکس در سایت بارگذاری شوند، ترتیب اثری داده نخواهد شد.</li>
<li>
 همه‌ی کدهای پیوست گزارش بایستی قابلیت اجرای مجدد داشته باشند. در صورتی که برای اجرا مجدد آن‌ها نیاز به تنظیمات خاصی می‌باشد، بایستی تنظیمات مورد نیاز را نیز در گزارش خود ذکر کنید.  دقت کنید که  تمامی کدها باید توسط شما اجرا شده باشند و نتایج اجرا در فایل کدهای ارسالی مشخص باشد. به کدهایی که نتایج اجرای آن‌ها در فایل ارسالی مشخص نباشد نمره‌ای تعلق نمی‌گیرد.
</li>
<li>توجه کنید این تمرین باید به صورت تک‌نفره انجام شود و پاسخ‌های ارائه شده باید نتیجه فعالیت فرد نویسنده باشد (همفکری و به اتفاق هم نوشتن تمرین نیز ممنوع است). در صورت مشاهده
 تشابه به همه افراد مشارکت‌کننده، نمره تمرین صفر و به استاد گزارش می‌گردد.
 </li>

 <li>
لطفاً تمامی پاسخ‌های متنی خود را با <b>فونت وزیر (Vazir)</b> و به‌صورت <b>راست‌چین</b> بنویسید.  
از استفاده از فونت‌های پیش‌فرض خودداری کنید تا ظاهر نوت‌بوک شما یک‌دست و خوانا باشد.  
در بخش‌های تشریحی، سعی کنید پاسخ‌ها را کامل، منسجم و با رعایت نگارش فارسی بنویسید.  
همچنین، به چینش تمیز سلول‌ها و اجرای درست کدها توجه کنید تا تمرین شما با فرمت خواسته‌شده و استاندارد ارائه شود.
</li>
 <li>برای مطالعه بیشتر درباره‌ی فرمت Markdown می‌توانید از <a href="https://github.com/tajaddini/Persian-Markdown/blob/master/learn-MD.md">این لینک</a> مطالعه کنید.
 </li>
 </ul>
    

 </div>
